demo/memnet/external_memory.py

```diff
@@ -0,0 +1,119 @@
+                                              initial_mean=0.))
+        self.is_test = is_test
+
+    def read(self, read_key):
+        cosine_similarity_read = cos_sim(read_key, self.external_memory, scale=self.scale, size=self.mem_slot_size)
+        norm_cosine_similarity_read = mixed_layer(input=
+                                                  identity_projection(cosine_similarity_read),
+                                                  bias_attr = False,
+                                                  act = SoftmaxActivation(),
+                                                  size = self.mem_slot_size,
+                                                  name='read_weight')
+
+        memory_read = linear_comb_layer(weights=norm_cosine_similarity_read, 
+                                        vectors=self.external_memory,
+                                        size=self.mem_fea_size, name='read_content')
+
+        if self.is_test:
+            print_layer(input=[norm_cosine_similarity_read, memory_read])
+
+        return memory_read
+
+    def write(self, write_key):
+        cosine_similarity_write = cos_sim(write_key, self.external_memory, 
+                                          scale=self.scale, size=self.mem_slot_size)
+        norm_cosine_similarity_write = mixed_layer(input=
+                                                   identity_projection(cosine_similarity_write),
+                                                   bias_attr = False,
+                                                   act = SoftmaxActivation(),
+                                                   size = self.mem_slot_size,
+                                                   name='write_weight')
+        if self.is_test:
+            print_layer(input=[norm_cosine_similarity_write])
+    
+        add_vec = mixed_layer(input = full_matrix_projection(write_key),
+                              bias_attr = None,
+                              act = SoftmaxActivation(),
+                              size = self.mem_fea_size,
+                              name='add_vector')
+
+        erase_vec = self.MakeConstantVector(self.mem_fea_size, 1.0, write_key)
+
+
+        if self.is_test:
+            print_layer(input=[erase_vec])
+            print_layer(input=[add_vec])
+
+        out_prod = out_prod_layer(norm_cosine_similarity_write, erase_vec, name="outer")
+
+        memory_remove = mixed_layer(input=dotmul_operator(a=self.external_memory, b=out_prod))
+
+        memory_remove_neg = slope_intercept_layer(input=memory_remove, slope=-1.0, intercept=0)
+
+        # memory_updated = memory_mat - memory_remove  = memory_mat + memory_remove_neg
+        memory_removed = mixed_layer(input = [identity_projection(input=self.external_memory),
+                                              identity_projection(input=memory_remove_neg)],
+                                     bias_attr = False,
+                                     act = LinearActivation())
```

Comment:
78 and 81 can be combinded as written as: memory_removed = self.external_memory - memory_remove.
See https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/tests/configs/math_ops.py